Prevents delegate role name as top-level role name

kairoaraujo opened the pull request:
This commit adds the validation in the ``metadata.Delegations``
to prevent that one of the delegate role names given is a top-level
role name.

A ``ValueError`` will be raised if one of the roles names in the
the list given to as delegated contains the role name as one of the
top-level roles.

The _mirrors role_ is not covered in this PR.

Signed-off-by: Kairo de Araujo <kdearaujo@vmware.com>

Please fill in the fields below to submit a pull request.  The more information
that is provided, the better.

Fixes #1558 

**Description of the changes being introduced by the pull request**:

**Please verify and check that the pull request fulfills the following
requirements**:

- [X] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [X] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Prevents delegate role name as top-level role name

This commit adds the validation in the ``metadata.Delegations``
to prevent that one of the delegate role names given is a top-level
role name.

A ``ValueError`` will be raised if one of the roles names in the
list given to as delegated contains the role name as one of the
top-level roles.

Signed-off-by: Kairo de Araujo <kdearaujo@vmware.com>

MVrachev commented on tests/test_metadata_serialization.py:
```diff
@@ -360,6 +360,43 @@ def test_delegation_serialization(self, test_case_data: str):
+        "using root as delegate role name": '{"keys": { \
+                "keyid1" : {"keytype": "rsa", "scheme": "rsassa-pss-sha256", "keyval": {"public": "foo"}}}, \
+            "roles": [ \
+                {"keyids": ["keyid"], "name": "root", "terminating": true, "paths": ["fn1"], "threshold": 3}] \
```
```suggestion
                {"keyids": ["keyid1"], "name": "root", "terminating": true, "paths": ["fn1"], "threshold": 3}] \
```

Please note that the `keyids` used in `Targets.Delegations.DelegationRole.keys` should then be used inside `Targets.Delegations.DelegationRole.roles`.
See an example in our hardcoded targets file [here](https://github.com/theupdateframework/python-tuf/blob/develop/tests/repository_data/repository/metadata/targets.json).

Please make that change in the other test cases as well.

kairoaraujo replied:
I see I messed up it.

MVrachev commented on tests/test_metadata_serialization.py:
```diff
@@ -360,6 +360,43 @@ def test_delegation_serialization(self, test_case_data: str):
         delegation = Delegations.from_dict(copy.deepcopy(case_dict))
         self.assertDictEqual(case_dict, delegation.to_dict())
 
+    valid_delegations: utils.DataSet = {
```
Can you please use the dataset defined at line 321 https://github.com/theupdateframework/python-tuf/blob/acb201d6cc685c2560f1000faf47e75db73df7fb/tests/test_metadata_serialization.py#L321
and the test function defined in 336 https://github.com/theupdateframework/python-tuf/blob/acb201d6cc685c2560f1000faf47e75db73df7fb/tests/test_metadata_serialization.py#L336
instead of defining a `valid_delegations` and a new test function?

We aim to group all valid tests related to a specific class in one dataset and test function.
Also, the name `valid_delegations` is not correct.
Those are actually cases of `invalid_delegations` as when you call `Delegations.from_dict(copy.deepcopy(case_dict))` it causes `ValueError` as you catch it inside your the test function.

kairoaraujo replied:
I have a concern about how to make sure that the ``ValueError`` is the correct unit we want.
I mean, we test the ``ValueError`` raised, but we don't check if the message comes from the part we want to cover.
Unless we don't want to go deep in this level, it is ok.

MVrachev replied:
I think this is a more general question that probably is worth adressing.
I think for sure we want to have all invalid tests for `Delegations` in one single place, but I understand what you mean.
Maybe we can check the error message and type inside the with?
What do you think @jku?

kairoaraujo replied:
Yes, indeed. If agreed in checking the type and error message, we can address it in another issue.

MVrachev commented on tests/test_metadata_serialization.py:
```diff
@@ -360,6 +360,43 @@ def test_delegation_serialization(self, test_case_data: str):
+        "using timestamp as delegate role name": '{"keys": { \
+                "keyid1" : {"keytype": "rsa", "scheme": "rsassa-pss-sha256", "keyval": {"public": "foo"}}}, \
+            "roles": [ \
+                {"keyids": ["keyid"], "name": "root", "terminating": true, "paths": ["fn1"], "threshold": 3}] \
```
```suggestion
                {"keyids": ["keyid"], "name": "timestamp", "terminating": true, "paths": ["fn1"], "threshold": 3}] \
```

MVrachev commented on tests/test_metadata_serialization.py:
```diff
@@ -360,6 +360,43 @@ def test_delegation_serialization(self, test_case_data: str):
+                "keyid1" : {"keytype": "rsa", "scheme": "rsassa-pss-sha256", "keyval": {"public": "foo"}}, \
+                "keyid2" : {"keytype": "ed25519", "scheme": "ed25519", "keyval": {"public": "bar"}}}, \
+            "roles": [ \
+                {"keyids": ["keyid"], "name": "b", "terminating": true, "paths": ["fn1"], "threshold": 3}, \
```
```suggestion
                {"keyids": ["keyid1"], "name": "b", "terminating": true, "paths": ["fn1"], "threshold": 3}, \
```

MVrachev commented on tuf/api/metadata.py:
```diff
@@ -1213,6 +1213,12 @@ def __init__(
     ):
         self.keys = keys
+        if [role for role in set(roles) if role in TOP_LEVEL_ROLE_NAMES]:
+            raise ValueError(
+                "A delegated role can not use top-level role names ("
+                f"{', '.join(TOP_LEVEL_ROLE_NAMES)})"
+            )
```
Can you use a shorter exception message?
That's something I have been guilty of on numerous occasions and @jku asked me to refine my messages. :D 
Maybe something like
``` suggestion
            raise ValueError("Delegated roles cannot use top-level role names")
```
An example of a similar message for Root roles where they should be top-level metadata names:
https://github.com/theupdateframework/python-tuf/blob/acb201d6cc685c2560f1000faf47e75db73df7fb/tuf/api/metadata.py#L733

kairoaraujo replied:
Sounds good. What about only showing the role conflicting, such as:
https://github.com/theupdateframework/python-tuf/blob/acb201d6cc685c2560f1000faf47e75db73df7fb/tuf/api/metadata.py#L1231

MVrachev replied:
I don't think it's absolutely correct to say that this role is a duplicate.
Yes, we are using the role name for our top-level roles, but it can be understood as if there is another role with the same name inside `Targets.Delegations` which is untrue.
Here the focus is that we don't want delegation role names to be one of the top-level metadata roles as this could lead to unexpected behaviors.

MVrachev reviewed (changes requested):
Thank you @kairoaraujo for your pr and effort!
There are a couple of things to fix, but it's a great start.

Please note that at the current stage `mirrors` won't be implemented in the new implementation. For more reference read the second point from issue #1317.

Commit:
Fix reuse datasets/tests, better keyids, standards

- Reuse the dataset and the existing tests
- Fix the keyids in the tests datasets to be aligned
- Fix the ``ValueError`` message aligned to the existent messages

Signed-off-by: Kairo de Araujo <kdearaujo@vmware.com>

sechkova commented:
The [issue discussion](https://github.com/theupdateframework/python-tuf/issues/1558#issuecomment-924846015) mentions also preventing delegated roles from being empty strings. Do you mind adding this check as part of this PR? Sorry if it's been done already somewhere and I've missed that.

Commit:
Ccheck empty strings, remove comprehensive lists

- Add the check for empty strings in the Delegate Role name
- Remove the comprehensive lists to make the code more readable
- Remove the test for empty file name from
``test_updater_with_simulator``

Signed-off-by: Kairo de Araujo <kdearaujo@vmware.com>

MVrachev commented on tuf/api/metadata.py:
```diff
@@ -1213,8 +1213,12 @@ def __init__(
-        if [role for role in set(roles) if role in TOP_LEVEL_ROLE_NAMES]:
-            raise ValueError("Delegated roles cannot use top-level role names")
+
+        for role in set(roles):
+            if not role or role in TOP_LEVEL_ROLE_NAMES:
+                raise ValueError(
+                    "Delegated roles cannot be empty or use top-level role names"
+                )
```
You still can write the check as a one-line like this:
```suggestion
        if any(not role or role in TOP_LEVEL_ROLE_NAMES for role in set(roles)):
            raise ValueError(
                "Delegated roles cannot be empty or use top-level role names"
            )
```
and I personally prefer it.
@sechkova what do you think?

kairoaraujo replied:
Yes, I had a small talk with @jku, and we came out about making it more readable. I'm open about any option 🙂

sechkova replied:
Thanks for adding the check, I agree that separate lines are more readable.

jku replied:
set() does not seem necessary?

MVrachev commented:
@sechkova can you approve running the CI workflows as GitHub says that:
`First-time contributors need a maintainer to approve running workflows`.

sechkova reviewed (commented):
I don't know if @MVrachev has any more comments, looks ok to me.

MVrachev commented:
Thanks for asking.
I don't have any other comments.
LGTM!

jku reviewed (commented):
LGTM, but there does seem to be an unnecessary set construction in there

Commit:
Remove the unnecessary set()

The set() is not required in the OrderedDict.

Signed-off-by: Kairo de Araujo <kdearaujo@vmware.com>

jku commented:
Thanks!
